src/raft_server.c

```diff
@@ -919,30 +932,24 @@ int raft_apply_entry(raft_server_t* me_)
 
     int node_id = me->cb.log_get_node_id(me_, raft_get_udata(me_), ety, log_idx);
     raft_node_t* node = raft_get_node(me_, node_id);
+    assert(node);
 
     switch (ety->type) {
         case RAFT_LOGTYPE_ADD_NODE:
-            raft_node_set_addition_committed(node, 1);
-            raft_node_set_voting_committed(node, 1);
             /* Membership Change: confirm connection with cluster */
             raft_node_set_has_sufficient_logs(node);
             if (node_id == raft_get_nodeid(me_))
                 me->connected = RAFT_NODE_STATUS_CONNECTED;
             break;
-        case RAFT_LOGTYPE_ADD_NONVOTING_NODE:
-            raft_node_set_addition_committed(node, 1);
-            break;
-        case RAFT_LOGTYPE_DEMOTE_NODE:
-            if (node)
-                raft_node_set_voting_committed(node, 0);
-            break;
         case RAFT_LOGTYPE_REMOVE_NODE:
-            if (node)
+            /* Do not remove node if there are pending entries that affect it */
+            if (raft_node_get_offered_idx(node) == log_idx)
```

Comment:
What cases require this?

Reply:
I wonder if many of the difficulties come from the following case:

1. Offer entry "remove node 3".
2. Offer entry "add nonvoting node 3".
3. Commit the entry offered in step 1.

Would it be fair to say the "add nonvoting" entry must wait for the "remove" entry to be committed, as they conflict on the same node ID?

In each raft_node_t object (and its user data), we maintain states like next_idx, match_idx, offered_idx and (on leaders) the InstallSnapshot progress. It would be easier to understand if all these get reset for the entry "add nonvoting" above, it seems.

Reply:
P.S., I think the Raft discussions all assume a new replica starts with fresh state. That is, we consider the "node 3" is first removed, then another fresh node is added but using ID 3 that has been used previously.

Reply:
OK, I'll land this. We can keep thinking about clarifications, correctness arguments, and improvements.